core/string/ustring.cpp

```diff
@@ -93,118 +91,61 @@ bool select_word(const String &p_s, int p_col, int &r_beg, int &r_end) {
 	if (length() == 0) {
-		return p_right.length() != 0;
+		return p_other.length() != 0;
 	}
 
-	return is_str_less(get_data(), p_right.get_data());
-}
-
-Char16String &Char16String::operator+=(char16_t p_char) {
-	const int lhs_len = length();
-	resize(lhs_len + 2);
-
-	char16_t *dst = ptrw();
-	dst[lhs_len] = p_char;
-	dst[lhs_len + 1] = 0;
-
-	return *this;
+	return is_str_less(get_data(), p_other.get_data());
 }
 
-void Char16String::operator=(const char16_t *p_cstr) {
-	copy_from(p_cstr);
-}
-
-const char16_t *Char16String::get_data() const {
-	if (size()) {
-		return &operator[](0);
-	} else {
-		return u"";
-	}
-}
-
-void Char16String::copy_from(const char16_t *p_cstr) {
-	if (!p_cstr) {
-		resize(0);
-		return;
-	}
-
-	const char16_t *s = p_cstr;
-	for (; *s; s++) {
-	}
-	size_t len = s - p_cstr;
-
-	if (len == 0) {
-		resize(0);
-		return;
+template <typename T>
+bool CharStringT<T>::operator==(const CharStringT<T> &p_other) const {
+	if (length() != p_other.length()) {
+		return false;
 	}
-
-	Error err = resize(++len); // include terminating null char
-
-	ERR_FAIL_COND_MSG(err != OK, "Failed to copy char16_t string.");
-
-	memcpy(ptrw(), p_cstr, len * sizeof(char16_t));
-}
-
-/*************************************************************************/
-/*  CharString                                                           */
-/*************************************************************************/
-
-bool CharString::operator<(const CharString &p_right) const {
 	if (length() == 0) {
-		return p_right.length() != 0;
+		return true;
 	}
 
-	return is_str_less(get_data(), p_right.get_data());
-}
+	size_t len = length();
+	const T *src = get_data();
+	const T *dst = p_other.get_data();
 
-bool CharString::operator==(const CharString &p_right) const {
-	if (length() == 0) {
-		// True if both have length 0, false if only p_right has a length
-		return p_right.length() == 0;
-	} else if (p_right.length() == 0) {
-		// False due to unequal length
-		return false;
+	// Compare char by char
+	for (size_t i = 0; i < len; i++) {
+		if (src[i] != dst[i]) {
+			return false;
+		}
 	}
 
-	return strcmp(ptr(), p_right.ptr()) == 0;
+	return true;
 }
 
-CharString &CharString::operator+=(char p_char) {
+template <typename T>
+CharStringT<T> &CharStringT<T>::operator+=(T p_char) {
 	const int lhs_len = length();
 	resize(lhs_len + 2);
 
-	char *dst = ptrw();
+	T *dst = ptrw();
 	dst[lhs_len] = p_char;
 	dst[lhs_len + 1] = 0;
 
 	return *this;
 }
 
-void CharString::operator=(const char *p_cstr) {
-	copy_from(p_cstr);
-}
-
-const char *CharString::get_data() const {
-	if (size()) {
-		return &operator[](0);
-	} else {
-		return "";
-	}
-}
-
-void CharString::copy_from(const char *p_cstr) {
+template <typename T>
+void CharStringT<T>::copy_from(const T *p_cstr) {
 	if (!p_cstr) {
 		resize(0);
 		return;
 	}
 
-	size_t len = strlen(p_cstr);
+	const T *s = p_cstr;
+	for (; *s; s++) {
+	} // Iterate until null char reached.
+	size_t len = s - p_cstr;
```

Comment:
```suggestion
	size_t len = strlen(p_cstr);
```

There are `strlen` overloads now for the other char types :)